dss/api/collections.py

```diff
@@ -0,0 +1,190 @@
+        raise DSSException(requests.codes.conflict,
+                           f"collection_tombstone_already_exists",
+                           f"collection tombstone with UUID {uuid} already exists")
+    status_code = requests.codes.ok
+    response_body = dict()  # type: dict
+
+    return jsonify(response_body), status_code
+
+@functools.lru_cache(maxsize=64)
+def get_json_metadata(entity_type: str, uuid: str, version: str, replica: Replica, blobstore_handle: BlobStore):
+    try:
+        key = "{}s/{}.{}".format(entity_type, uuid, version)
+        # TODO: verify that file is a metadata file
+        size = blobstore_handle.get_size(replica.bucket, key)
+        if size > MAX_METADATA_SIZE:
+            raise DSSException(422, "invalid_link",
+                               "The file UUID {} refers to a file that is too large to process".format(uuid))
+        return json.loads(blobstore_handle.get(
+            replica.bucket,
+            "{}s/{}.{}".format(entity_type, uuid, version)))
+    except BlobNotFoundError as ex:
+        raise DSSException(404, "invalid_link", "Could not find file for UUID {}".format(uuid))
+
+def resolve_content_item(replica: Replica, blobstore_handle: BlobStore, item: dict):
+    try:
+        if item["type"] in {"file", "bundle", "collection"}:
+            item_metadata = get_json_metadata(item["type"], item["uuid"], item["version"], replica, blobstore_handle)
+        else:
+            item_metadata = get_json_metadata("file", item["uuid"], item["version"], replica, blobstore_handle)
+            if "fragment" not in item:
+                raise Exception('The "fragment" field is required in collection elements '
+                                'other than files, bundles, and collections')
+            blob_path = "blobs/" + ".".join((
+                item_metadata[FileMetadata.SHA256],
+                item_metadata[FileMetadata.SHA1],
+                item_metadata[FileMetadata.S3_ETAG],
+                item_metadata[FileMetadata.CRC32C],
+            ))
+            # check that item is marked as metadata, is json, and is less than max size
+            item_doc = json.loads(blobstore_handle.get(replica.bucket, blob_path))
+            item_content = jsonpointer.resolve_pointer(item_doc, item["fragment"])
+            return item_content
+    except DSSException:
+        raise
+    except Exception as e:
+        raise DSSException(
+            422,
+            "invalid_link",
+            'Error while parsing the link "{}": {}: {}'.format(item, type(e).__name__, e)
+        )
+
+def verify_collection(contents: List[dict], replica: Replica, blobstore_handle: BlobStore):
+    """
+    Given user-supplied collection contents that pass schema validation, resolve all entities in the collection and
+    verify they exist.
+    """
+    executor = ThreadPoolExecutor(max_workers=8)
+    for result in executor.map(partial(resolve_content_item, replica, blobstore_handle), contents):
```

Comment:
Question: This will raise and abort the loop on the first validation error leaving the TPE running until it is GCed, right?

Suggestion: I'd be more comfortable with partitioning the collection into batches of 100 items, running each batch separately through the TPE (via .submit()), exhausting all results by collecting the futures and populating the response with all errors in a batch. Or maybe accumulate errors until you have more than a certain threshold. Also timeouts should be handled more gracefully. This implementation seems overly naìve, but I'm probably overlooking something here which is why I've marked this as a suggestion.

Reply:
The implementation is deliberately simple as a first cut. Yes, the cleanup of the threadpool will need to be tightened up in the future to make sure no threads make it past the return from this frame. I don't think batching is needed. I've added a task in #1269.

Reply:
I'd be happy with a simplest possible implementation that demonstrably didn't leak threads.

Reply:
Updated as per the other comment.

Added a task in #1269 to make sure meaningful errors are thrown on timeout.